Scripts: Improve handling of check-licenses for optional dependencies

aduth opened the pull request:
## What?

Updates `check-licenses` script to resolve issues in some environments where optional dependencies may be "installed" as empty folders in `node_modules`, causing the script to fail because the package's `package.json` cannot be read.

Follow-up to #72996

## Why?

To fix the build failures on `trunk`, which currently have this issue.

Example: https://github.com/WordPress/gutenberg/actions/runs/19111516403/job/54609577907

## How?

Using [`--package-lock-only` option with `npm ls`](https://docs.npmjs.com/cli/v9/commands/npm-ls#package-lock-only) does two things:

- It ignores the file structure (i.e. empty folders) in `node_modules`, relying exclusively on what's in `package-lock.json`
- It includes the full package info as part of the result, including the `license` field we're interested in

An earlier iteration in f0dbfab28135f6fc6c8f6d84e4212e223099eecd was more targeted at optional dependencies and checking supported platform, but this felt unnecessarily complex.

## Testing Instructions

Verify that `npm run other:check-licenses` passes.

To validate that the intended behavior of license checking works, try editing a license to something invalid in `package-lock.json` and in the package's own license indicators (`package.json` `license` field, README references), particularly with some or all of the following conditions:

1. `devDependencies`
2. Transitive dependencies (i.e. dependencies of dependencies)
3. Optional dependencies

For example, I made the following revision and verified that this is flagged as an incompatible license:

```diff
diff --git a/package-lock.json b/package-lock.json
index 1f11eca92b..1509124ca8 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -43713,3 +43713,3 @@
 			"dev": true,
-			"license": "MIT",
+			"license": "MITbad",
 			"optional": true,
```

Plus manual edits to `node_modules/sass-embedded-darwin-arm64/package.json` to change `license` from `"MIT"` to `"MITbad"`.

```
npm exec wp-scripts check-licenses -- --dev
[...]
 ERROR  Module sass-embedded-darwin-arm64 has an incompatible license 'MITbad'.
```

aduth committed:
Scripts: Skip optional dependencies for check-licenses on unsupported platform

aduth committed:
Scripts: Use known package info to extract license

t-hamano reviewed (approved):
LGTM! It worked as expected on Windows OS.

<img width="1447" height="584" alt="Image" src="https://github.com/user-attachments/assets/7ad397d6-6a5b-4f52-87a6-d623f531fe12" />

aduth committed:
Add fallback to read packageInfo from package directory

aduth commented:
> I think I might add back the original logic as a fallback if the license isn't available from `package-lock.json` fields.

Added in 53413fd.

aduth committed:
Scripts: Skip optional dependencies for check-licenses on unsupported platform

aduth committed:
Consider packages from filesystem

aduth committed:
Build up optional dependencies from tree

When not using --package-lock-only, the package information doesn't include whether the package itself is optional, but we can build up a set of known optional dependencies based on what top-level dependencies define as their own optional dependencies

aduth commented:
Hm, I'm actually not sure about this solution:

- Just because a package is defined as an optional dependency by one package doesn't mean that it should be considered optional for the entire tree
- I'm similarly confused why we even get to the point where we're checking licenses on optional dependencies to begin with, because [the logic only recurses into `dependencies`](https://github.com/WordPress/gutenberg/blob/70f48bfe018d393b738bc3e3c88046378c775abe/packages/scripts/utils/license.js#L345-L347), not `optionalDependencies`.

sirreal commented:
Are you familiar with `npm query` and have you tried it?

- https://docs.npmjs.com/cli/v11/using-npm/dependency-selectors#dependency-type-selectors
- https://docs.npmjs.com/cli/v11/commands/npm-query

Something like this might list the desired packages:

`npm query '.workspace :not(.optional, .dev)'`

**Edit:**

Check licenses uses it here:

https://github.com/WordPress/gutenberg/blob/5db7f248cf331fec09572b05ab03cce47a09cb89/bin/check-licenses.mjs#L26-L40

[The docs have also lead me to `@npmcli/arborist` which I believe powers the query command and is likely a powerful npm tree querying tool.](https://www.npmjs.com/package/@npmcli/arborist)

jsnajdr commented:
If we want to be future-proof, it's good to known that pnpm's `pnpm-lock.yml` file doesn't contain the license information at all, and `pnpm list` cannot list it either. We shouldn't rely on NPM features that will prevent a seamless switch in the future.

aduth committed:
Add inline code comments

aduth committed:
Use npm query to produce list of dependencies

aduth committed:
Update comment referencing npm ls

aduth committed:
Update project-specific license checker to use npm query

aduth commented:
Marking this ready for review. I think this approach using `npm query` could work and be more efficient at leveraging npm to do the work of filtering out packages with permissible licenses as a first pass.

aduth commented:
I updated the testing instructions in the original comment. `sass-embedded-darwin-arm64` is a pretty good one to try because it only requires updating `node_modules/sass-embedded-darwin-arm64/package.json` in addition to `package-lock.json` (it doesn't include license details anywhere else in the package).

aduth commented:
Verifying the code changes in `bin/check-licenses.mjs` is similar, but picking a (production) dependency of one of the packages that's shipped as a script.

I picked `deepmerge` (dependency of `@wordpress/block-editor`)

- Modify `license` value for `node_modules/deepmerge` in `package-lock.json`
- Modify `license` value in `node_modules/deepmerge/package.json`
- `rm node_modules/deepmerge/license.txt node_modules/deepmerge/readme.md`

```
node ./bin/check-licenses.mjs
 ERROR  Module deepmerge has an incompatible license 'MITbad'
```

sirreal reviewed (approved):
Looks good and works as expected in my testing 👍
